### ISSUE TITLE
"System.ArgumentException: The path is not of a legal form." - it's not?

### ISSUE THREAD

**Issue by ziipzaaapM16A4** [sandrohanea/whisper.net#253]
```
1/7/2024 5:04:21 AM.342] LSPD First Response: [DispatchLLM] ERROR: System.ArgumentException: The path is not of a legal form.
[11/7/2024 5:04:21 AM.346] at System.IO.Path.LegacyNormalizePath(String path, Boolean fullCheck, Int32 maxPathLength, Boolean expandShortPaths)
[11/7/2024 5:04:21 AM.347] at System.IO.Path.InternalGetDirectoryName(String path)
[11/7/2024 5:04:21 AM.347] at Whisper.net.LibraryLoader.NativeLibraryLoader.<GetRuntimePaths>d__4.MoveNext()
[11/7/2024 5:04:21 AM.348] at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
[11/7/2024 5:04:21 AM.349] at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
[11/7/2024 5:04:21 AM.349] at Whisper.net.LibraryLoader.NativeLibraryLoader.LoadLibraryComponent()
[11/7/2024 5:04:21 AM.350] at Whisper.net.LibraryLoader.NativeLibraryLoader.LoadNativeLibrary()
[11/7/2024 5:04:21 AM.351] at Whisper.net.WhisperFactory.<>c.<.cctor>b__10_0()
[11/7/2024 5:04:21 AM.352] at System.Lazy`1.CreateValue()
[11/7/2024 5:04:21 AM.352] at System.Lazy`1.LazyInitValue()
[11/7/2024 5:04:21 AM.353] at Whisper.net.WhisperFactory..ctor(IWhisperProcessorModelLoader loader, Boolean delayInit)
[11/7/2024 5:04:21 AM.354] at Whisper.net.WhisperFactory.FromPath(String path, Boolean delayInitialization)
[11/7/2024 5:04:21 AM.354] at Dispatch_LLM.SpeechToTextManager..ctor() in D:\#UserData\Dokumente 2 - TB\Development\Visual Studio\Repos\Dispatch LLM\SpeechToTextManager.cs:line 26

```
Code is:
```
var modelName = "ggml-base.bin";
whisperFactory = WhisperFactory.FromPath(modelName);     //Here it crashes
```

Wanted Path is
G:\SteamLibrary\steamapps\common\Grand Theft Auto V\ggml-base.bin

Work on my console program flawless. Even tho the console program path has more complex syntax. for example it uses # in its path.

C# DotNet Framework.    
    <PlatformTarget>x64</PlatformTarget>
    <LangVersion>11.0</LangVersion>
    
Did i miss something?

**Comment by sandrohanea** [comment]
Hello @ziipzaaapM16A4 ,

Based on the stacktrace, the problem is not with the path for the model, but with the path for the native library:
https://github.com/sandrohanea/whisper.net/blob/268934fd7fd724d615c32e3a5410c201cc236e4d/Whisper.net/LibraryLoader/NativeLibraryLoader.cs#L196

Can you, please, try to see which one if these are failing for your system?

```
  var assemblyLocation = typeof(NativeLibraryLoader).Assembly.Location;
        var environmentAppStartLocation = Environment.GetCommandLineArgs()[0];
        // NetFramework and Mono will crash if we try to get the directory of an empty string.
        var assemblySearchPaths = new[]
            {
                string.IsNullOrWhiteSpace(RuntimeOptions.Instance.LibraryPath) ? null : Path.GetDirectoryName(RuntimeOptions.Instance.LibraryPath),
                AppDomain.CurrentDomain.RelativeSearchPath,
                AppDomain.CurrentDomain.BaseDirectory,
                string.IsNullOrWhiteSpace(assemblyLocation) ? null : Path.GetDirectoryName(assemblyLocation),
                string.IsNullOrWhiteSpace(environmentAppStartLocation) ? null : Path.GetDirectoryName(environmentAppStartLocation),
            }.Where(it => !string.IsNullOrEmpty(it)).Distinct();
```

It would be great to also have the actual path that was generated in order to detect and skip it.

I expect it is either `assemblyLocation` or `environmentAppStartLocation `:
```
  var assemblyLocation = typeof(NativeLibraryLoader).Assembly.Location;
  var environmentAppStartLocation = Environment.GetCommandLineArgs()[0];
  Path.GetDirectoryName(assemblyLocation);
  Path.GetDirectoryName(environmentAppStartLocation);
```

**Comment by sandrohanea** [comment]
Added https://github.com/sandrohanea/whisper.net/commit/4f6e22e42d1a39dbaca9289baecac3ac85bd9cef to ensure that all frameworks will allow that DirectoryName to be parsed. It will be available in the next release.

**Comment by ziipzaaapM16A4** [comment]
> Can you, please, try to see which one if these are failing for your system?
> 
> ```
>   var assemblyLocation = typeof(NativeLibraryLoader).Assembly.Location;
>         var environmentAppStartLocation = Environment.GetCommandLineArgs()[0];
>         // NetFramework and Mono will crash if we try to get the directory of an empty string.
>         var assemblySearchPaths = new[]
>             {
>                 string.IsNullOrWhiteSpace(RuntimeOptions.Instance.LibraryPath) ? null : Path.GetDirectoryName(RuntimeOptions.Instance.LibraryPath),
>                 AppDomain.CurrentDomain.RelativeSearchPath,
>                 AppDomain.CurrentDomain.BaseDirectory,
>                 string.IsNullOrWhiteSpace(assemblyLocation) ? null : Path.GetDirectoryName(assemblyLocation),
>                 string.IsNullOrWhiteSpace(environmentAppStartLocation) ? null : Path.GetDirectoryName(environmentAppStartLocation),
>             }.Where(it => !string.IsNullOrEmpty(it)).Distinct();
> ```

Its says
RuntimeOptions.Instance.LibraryPath => LibraryPath is not existent => Compiler Error

The other code:

```
            var assemblyLocation = typeof(NativeLibraryLoader).Assembly.Location;
            var environmentAppStartLocation = Environment.GetCommandLineArgs()[0];
            RONFuncs.LogTrivial("sandrohanea code sample: " + assemblyLocation.ToString());
            RONFuncs.LogTrivial("sandrohanea code sample: " + environmentAppStartLocation.ToString());
            RONFuncs.LogTrivial("sandrohanea code sample: " + Path.GetDirectoryName(assemblyLocation));
            RONFuncs.LogTrivial("sandrohanea code sample: " + Path.GetDirectoryName(environmentAppStartLocation));
```

 results:

```
[11/8/2024 8:04:02 PM.468] LSPD First Response: sandrohanea code sample:
[11/8/2024 8:04:02 PM.470] LSPD First Response: sandrohanea code sample: G:\SteamLibrary\steamapps\common\Grand Theft Auto V\GTA5.exe
[11/8/2024 8:04:02 PM.472] LSPD First Response: [DispatchLLM] ERROR: System.ArgumentException: The path is not of a legal form.
[11/8/2024 8:04:02 PM.473] at System.IO.Path.LegacyNormalizePath(String path, Boolean fullCheck, Int32 maxPathLength, Boolean expandShortPaths)
[11/8/2024 8:04:02 PM.474] at System.IO.Path.InternalGetDirectoryName(String path)
[11/8/2024 8:04:02 PM.474] at Dispatch_LLM.SpeechToTextManager..ctor() in D:\#UserData\Dokumente 2 - TB\Development\Visual Studio\Repos\Dispatch LLM\SpeechToTextManager.cs:line 37
[11/8/2024 8:04:02 PM.475] at Dispatch_LLM.Main.<>c.<StartPlugin>b__4_0() in D:\#UserData\Dokumente 2 - TB\Development\Visual Studio\Repos\Dispatch LLM\Main.cs:line 58
```

**Comment by sandrohanea** [comment]
Got it, thanks!
Indeed, as the "environmentAppStartLocation" is inside the gta5 directory, it is causing issues for detecting the runtime dlls as it crashes before that. The commit with the try-catch above should help in this case.

**Comment by ziipzaaapM16A4** [comment]
but this crashes anyways right? so i will be unable to use it. 

or did i not understand this?

**Comment by sandrohanea** [comment]
Even if manually setting the path with `RuntimeOptions.Instance.SetLibraryPath(...)` that version would still crash for your setup as all the paths are evaluated including `environmentAppStartLocation`.

The fix was released in 1.7.2 yesteday, and you should be able to use that version that handles the exceptions for netframework setup with invalid path on one of the options.
